File: python/browser/parser.py
class Text:
    def __init__(self, text, parent):
        self.text = text
        self.children = []
        self.parent = parent

    def __repr__(self):
        return repr(self.text)


class Element:
    def __init__(self, tag, attributes, parent):
        self.tag = tag
        self.children = []
        self.parent = parent
        self.attributes = attributes

    def __repr__(self):
        return "<" + self.tag + ">"


class HTMLParser:
    SELF_CLOSING_TAGS = [
        "area", "base", "br", "col", "embed", "hr", "img", "input",
        "link", "meta", "param", "source", "track", "wbr",
    ]

    HEAD_TAGS = [
        "base", "basefont", "bgsound", "noscript",
        "link", "meta", "title", "style", "script",
    ]

    def __init__(self, body):
        self.body = body
        self.unfinished = []

    def parse(self):
        text = ""
        in_tag = False
        for c in self.body:
            if c == "<":
                in_tag = True
                if text:
                    self.add_text(text)
                text = ""
            elif c == ">":
                in_tag = False
                self.add_tag(text)
                text = ""
            else:
                text += c
        if not in_tag and text:
            self.add_text(text)
        return self.finish()

    def add_text(self, text):
        if text.isspace():
            return
        self.implicit_tags(None)
        parent = self.unfinished[-1]
        node = Text(text, parent)
        parent.children.append(node)

    def get_attributes(self, text):
        parts = text.split()
        tag = parts[0].casefold()
        attributes = {}
        for attrpair in parts[1:]:
            if "=" in attrpair:
                key, value = attrpair.split("=", 1)
                attributes[key.casefold()] = value
                if len(value) > 2 and value[0] in ["'", "\""]:
                    value = value[1:-1]
            else:
                attributes[attrpair.casefold()] = ""
        return tag, attributes

    def add_tag(self, tag):
        tag, attributes = self.get_attributes(tag)
        if tag.startswith("!"):
            return
        self.implicit_tags(tag)
        if tag.startswith("/"):
            if len(self.unfinished) == 1:
                return
            node = self.unfinished.pop()
            parent = self.unfinished[-1]
            parent.children.append(node)
        elif tag in self.SELF_CLOSING_TAGS:
            parent = self.unfinished[-1]
            node = Element(tag, attributes, parent)
            parent.children.append(node)
        else:
            parent = self.unfinished[-1] if self.unfinished else None
            node = Element(tag, attributes, parent)
            self.unfinished.append(node)
    
    def implicit_tags(self, tag):
        while True:
            open_tags = [node.tag for node in self.unfinished]
            if open_tags == [] and tag != "html":
                self.add_tag("html")
            elif open_tags == ["html"] and tag not in ["head", "body", "/html"]:
                if tag in self.HEAD_TAGS:
                    self.add_tag("head")
                else:
                    self.add_tag("body")
            elif open_tags == ["html", "head"] and tag not in ["/head"] + self.HEAD_TAGS:
                self.add_tag("/head")
            else:
                break

    def finish(self):
        if not self.unfinished:
            self.implicit_tags(None)
        while len(self.unfinished) > 1:
            node = self.unfinished.pop()
            parent = self.unfinished[-1]
            parent.children.append(node)
        return self.unfinished.pop()


def print_tree(node):
    print(tree_to_html(node))


def tree_to_html(node):
    if isinstance(node, Text):
        return node.text

    attributes = "".join(
        f' {key}="{value}"' if value else f" {key}"
        for key, value in node.attributes.items()
    )
    if node.tag in HTMLParser.SELF_CLOSING_TAGS:
        return f"<{node.tag}{attributes}>"

    children = "".join(tree_to_html(child) for child in node.children)
    return f"<{node.tag}{attributes}>{children}</{node.tag}>"


def extract_text(tokens):
    text = ""
    entity = ""
    in_entity = False
    in_whitespace = False

    def visit(node):
        nonlocal text, entity, in_entity, in_whitespace

        if isinstance(node, Element):
            normalized_tag = node.tag.strip().casefold()
            if normalized_tag in ["br", "br/"]:
                text = text.rstrip(" ")
                text += "\n"
                in_whitespace = False
                return

            for child in node.children:
                visit(child)

            if normalized_tag in ["div", "p"]:
                text = text.rstrip(" ")
                text += "\n"
                in_whitespace = False
            return

        for c in node.text:
            if in_entity:
                entity += c
                if entity == "&lt;":
                    text += "<"
                    entity = ""
                    in_entity = False
                    in_whitespace = False
                elif entity == "&gt;":
                    text += ">"
                    entity = ""
                    in_entity = False
                    in_whitespace = False
                elif c == ";":
                    text += entity
                    entity = ""
                    in_entity = False
                    in_whitespace = False
                continue

            if c == "&":
                entity = c
                in_entity = True
            elif c.isspace():
                if text and not in_whitespace:
                    text += " "
                in_whitespace = True
            else:
                text += c
                in_whitespace = False

    if isinstance(tokens, list):
        for token in tokens:
            visit(token)
    else:
        visit(tokens)

    return text.strip()

File: rust/src/parser.rs
use std::collections::HashMap;

#[derive(Clone, Debug)]
pub struct Text {
    pub text: String,
    pub children: Vec<Node>,
    pub style: HashMap<String, String>,
}

#[derive(Clone, Debug)]
pub struct Element {
    pub tag: String,
    pub attributes: HashMap<String, String>,
    pub children: Vec<Node>,
    pub style: HashMap<String, String>,
}

#[derive(Clone, Debug)]
pub enum Node {
    Text(Text),
    Element(Element),
}

pub struct HtmlParser {
    body: String,
    unfinished: Vec<Element>,
}

pub fn extract_text(node: &Node) -> String {
    let mut text = String::new();
    let mut entity = String::new();
    let mut in_entity = false;
    let mut in_whitespace = false;

    fn visit(
        node: &Node,
        text: &mut String,
        entity: &mut String,
        in_entity: &mut bool,
        in_whitespace: &mut bool,
    ) {
        match node {
            Node::Element(element) => {
                let normalized = element.tag.trim().to_ascii_lowercase();
                if matches!(normalized.as_str(), "br" | "br/") {
                    while text.ends_with(' ') {
                        text.pop();
                    }
                    text.push('\n');
                    *in_whitespace = false;
                    return;
                }

                for child in &element.children {
                    visit(child, text, entity, in_entity, in_whitespace);
                }

                if matches!(normalized.as_str(), "div" | "p") {
                    while text.ends_with(' ') {
                        text.pop();
                    }
                    text.push('\n');
                    *in_whitespace = false;
                }
            }
            Node::Text(text_node) => {
                for c in text_node.text.chars() {
                    if *in_entity {
                        entity.push(c);

                        if entity == "&lt;" {
                            text.push('<');
                            entity.clear();
                            *in_entity = false;
                            *in_whitespace = false;
                        } else if entity == "&gt;" {
                            text.push('>');
                            entity.clear();
                            *in_entity = false;
                            *in_whitespace = false;
                        } else if c == ';' {
                            text.push_str(entity);
                            entity.clear();
                            *in_entity = false;
                            *in_whitespace = false;
                        }
                        continue;
                    }

                    if c == '&' {
                        entity.push(c);
                        *in_entity = true;
                    } else if c.is_whitespace() {
                        if !text.is_empty() && !*in_whitespace {
                            text.push(' ');
                        }
                        *in_whitespace = true;
                    } else {
                        text.push(c);
                        *in_whitespace = false;
                    }
                }
            }
        }
    }

    visit(
        node,
        &mut text,
        &mut entity,
        &mut in_entity,
        &mut in_whitespace,
    );

    text.trim().to_string()
}

pub fn print_tree(node: &Node) {
    println!("{}", tree_to_html(node));
}

fn tree_to_html(node: &Node) -> String {
    match node {
        Node::Text(text) => text.text.clone(),
        Node::Element(element) => {
            let attributes = element
                .attributes
                .iter()
                .map(|(key, value)| {
                    if value.is_empty() {
                        format!(" {key}")
                    } else {
                        format!(" {key}=\"{value}\"")
                    }
                })
                .collect::<String>();

            if HtmlParser::SELF_CLOSING_TAGS.contains(&element.tag.as_str()) {
                format!("<{}{}>", element.tag, attributes)
            } else {
                let children = element.children.iter().map(tree_to_html).collect::<String>();
                if matches!(element.tag.as_str(), "html" | "head" | "body" | "div" | "p") {
                    format!(
                        "<{}{}>\n{}\n</{}>",
                        element.tag, attributes, children, element.tag
                    )
                } else {
                    format!("<{}{}>{}</{}>", element.tag, attributes, children, element.tag)
                }
            }
        }
    }
}

impl HtmlParser {
    const SELF_CLOSING_TAGS: [&'static str; 14] = [
        "area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "param",
        "source", "track", "wbr",
    ];
    const HEAD_TAGS: [&'static str; 9] = [
        "base",
        "basefont",
        "bgsound",
        "noscript",
        "link",
        "meta",
        "title",
        "style",
        "script",
    ];

    pub fn new(body: &str) -> Self {
        Self {
            body: body.to_owned(),
            unfinished: Vec::new(),
        }
    }

    pub fn parse(mut self) -> Node {
        let mut text = String::new();
        let mut in_tag = false;
        let body = self.body.clone();

        for c in body.chars() {
            if c == '<' {
                in_tag = true;
                if !text.is_empty() {
                    self.add_text(&text);
                }
                text.clear();
            } else if c == '>' {
                in_tag = false;
                self.add_tag(&text);
                text.clear();
            } else {
                text.push(c);
            }
        }

        if !in_tag && !text.is_empty() {
            self.add_text(&text);
        }
        self.finish()
    }

    fn add_text(&mut self, text: &str) {
        if text.trim().is_empty() {
            return;
        }
        self.implicit_tags(None);
        let parent = self
            .unfinished
            .last_mut()
            .expect("text node without an open parent element");
        parent.children.push(Node::Text(Text {
            text: text.to_owned(),
            children: Vec::new(),
            style: HashMap::new(),
        }));
    }

    fn get_attributes(text: &str) -> (String, HashMap<String, String>) {
        let mut parts = text.split_whitespace();
        let tag = parts.next().unwrap_or("").to_ascii_lowercase();
        let mut attributes = HashMap::new();

        for attr_pair in parts {
            if let Some((key, mut value)) = attr_pair.split_once('=') {
                if value.len() > 2 && matches!(value.chars().next(), Some('"') | Some('\'')) {
                    value = &value[1..value.len() - 1];
                }
                attributes.insert(key.to_ascii_lowercase(), value.to_owned());
            } else {
                attributes.insert(attr_pair.to_ascii_lowercase(), String::new());
            }
        }

        (tag, attributes)
    }

    fn add_tag(&mut self, raw_tag: &str) {
        let (tag, attributes) = Self::get_attributes(raw_tag);
        if tag.is_empty() || tag.starts_with('!') {
            return;
        }
        self.implicit_tags(Some(tag.as_str()));

        if tag.starts_with('/') {
            if self.unfinished.len() == 1 {
                return;
            }
            let node = self.unfinished.pop().expect("missing unfinished node");
            let parent = self.unfinished.last_mut().expect("missing parent node");
            parent.children.push(Node::Element(node));
        } else if Self::SELF_CLOSING_TAGS.contains(&tag.as_str()) {
            let parent = self
                .unfinished
                .last_mut()
                .expect("self-closing tag without an open parent element");
            parent.children.push(Node::Element(Element {
                tag,
                attributes,
                children: Vec::new(),
                style: HashMap::new(),
            }));
        } else {
            self.unfinished.push(Element {
                tag,
                attributes,
                children: Vec::new(),
                style: HashMap::new(),
            });
        }
    }

    fn implicit_tags(&mut self, tag: Option<&str>) {
        loop {
            let open_tags: Vec<&str> = self.unfinished.iter().map(|node| node.tag.as_str()).collect();
            if open_tags.is_empty() && tag != Some("html") {
                self.add_tag("html");
            } else if open_tags == ["html"] && !matches!(tag, Some("head" | "body" | "/html")) {
                if let Some(tag) = tag {
                    if Self::HEAD_TAGS.contains(&tag) {
                        self.add_tag("head");
                    } else {
                        self.add_tag("body");
                    }
                } else {
                    self.add_tag("body");
                }
            } else if open_tags == ["html", "head"]
                && !matches!(tag, Some("/head"))
                && !tag.is_some_and(|tag| Self::HEAD_TAGS.contains(&tag))
            {
                self.add_tag("/head");
            } else {
                break;
            }
        }
    }

    fn finish(&mut self) -> Node {
        if self.unfinished.is_empty() {
            self.implicit_tags(None);
        }
        while self.unfinished.len() > 1 {
            let node = self.unfinished.pop().expect("missing unfinished node");
            let parent = self.unfinished.last_mut().expect("missing parent node");
            parent.children.push(Node::Element(node));
        }
        Node::Element(self.unfinished.pop().expect("parser produced no root element"))
    }
}
